lm_eval/models/huggingface.py

```diff
@@ -403,11 +403,95 @@ def world_size(self):
     def tokenizer_name(self) -> str:
         return self.tokenizer.name_or_path.replace("/", "__")
 
-    @property
-    def chat_template(self) -> str:
-        if self.tokenizer.chat_template is not None:
-            return self.tokenizer.chat_template
-        return self.tokenizer.default_chat_template
+    def chat_template(self, chat_template: Optional[bool | str] = True) -> str:
+        """
+        Get the appropriate chat template for the model based on configuration and input.
+        This method determines, and returns the correct chat template, ensuring reproducibility.
+
+        The template selection logic is adapted from the Transformers library's `apply_chat_template`
+        method in the Tokenizer class. The original implementation can be found at:
+        https://github.com/huggingface/transformers/blob/fc35907f95459d7a6c5281dfadd680b6f7b620e3/src/transformers/tokenization_utils_base.py#L1687
+
+        This method ensures that the right template is chosen based on the following:
+        1. If the model's tokenizer has multiple templates:
+            a. Use the specified template if it exists in the dictionary.
+            b. Use the default template if no specific template is provided.
+            c. Raise an error if no default template exists and no specific template is specified.
+        2. If the model's tokenizer has a single template or no template:
+            a. Use the tokenizer's chat template if available.
+            b. Fall back to the default chat template if no tokenizer chat template exists.
+
+        Args:
+            chat_template (Optional[Union[bool, str]]): Specifies the chat template to use.
+                - If True (default), use the default template.
+                - If False or None, no template is applied (handled externally).
+                - If a string, use the template with the matching name.
+
+        Returns:
+            str: The selected chat template.
+        """
+        # This method should not be called if the chat_template argument is None or False
+        if chat_template is False or chat_template is None:
+            raise ValueError(
```

Comment:
Why not return None and print a warning instead?
For ex: "`model.chat_template` was called with the chat_template `False` or `None`. We're therefore not applying a chat template, make sure this is the behavior you needed.`

Reply:
This is great idea, updating